internals-js/src/definitions.ts

```diff
@@ -531,7 +548,7 @@ export abstract class SchemaElement<TOwnType extends SchemaElement<any, TParent>
 
   protected checkUpdate(addedElement?: { schema(): Schema | undefined }) {
```

Comment:
I don't think we ever pass something whose `schema()` method can be `undefined`.

But in fact, a suggestion would be to add to the parameter type that it should also have a `isAttached(): boolean` method so we avoid some `instanceof` later one. This does imply adding a `isAttached` to `BaseWrapperType` (but as its `schema` method can throw, I think it make sense to have a "can I call schema() safely" method in case it's needed) and rename `Element.hasParent` into `Element.isAttached` (that is, `hasParent` is just fine for `Element`, but it would be weird for `BaseWrapperType` because it doesn't expose a "parent", just a "schema", and so "isAttached" is a bit more general; a `hasSchema` naming would work to but I personally have a minor preference fort `isAttached`).

Pushed a commit with that suggestion at https://github.com/pcmanus/federation/commit/9aa17622194fd522782b810e84d6025487773a88 to clarify.